lib/plugins/extension/helper/extension.php

```diff
@@ -536,7 +588,8 @@ public function getDonationURL()
      */
     public function getTypes()
     {
-        if (!empty($this->remoteInfo['types'])) return $this->remoteInfo['types'];
+        if (!empty($this->remoteInfo['types']) && !$this->isIgnoredSource('remote'))
```

Comment:
Reading this over and over again, there are two things I notice.

1. The double negation of `!isIgnored` is awkward. Using a positive name `isSourceAllowed` would probably be better
2. I don't like all this repetition and wonder if it would be easier to unset the source somewhere, but I don't know how feasible this is.

Reply:
1. Good idea, will do that;
2. So not all these attributes use the same `isIgnoredSource` logic and pick the source in the same order of favor. Unsetting the source in `setExtension` with a new argument might be a better solution, will look into that.

And I agree this class needs a refactor. This mess of model/controller reusing the same class looks very, very bad to me now. If I had time somehow I might look into this.